support/default_output_validator.cpp

```diff
@@ -2,199 +2,379 @@
+		if (e < token.size() and not is_integer(token.substr(e + 1))) return false; // followed by an otional e<integer>
+		return true;
 	}
-	return output;
 }
 
-string quick_diff(const string& out, const string& ans) {
-	if(count(begin(ans), end(ans), '\n') <= 1 and count(begin(out), end(out), '\n') <= 1)
-		return crop_output("Got " + strip_newline(out) + " wanted " + strip_newline(ans));
-	return {};
+//============================================================================//
+// IO                                                                         //
+//============================================================================//
+std::string read_raw(std::istream& in) {
+	std::stringstream raw;
+	assert(in.good());
+	raw << in.rdbuf();
+	return raw.str();
 }
 
-pair<bool, string> default_output_validator(const string& ans_path, const string& feedback_dir) {
-	// Read answer.
-	string ans = [&] {
-		stringstream ans_stream;
-		ifstream f(ans_path);
-		ans_stream << f.rdbuf();
-		return ans_stream.str();
-	}();
-
-	// Read stdin.
-	string out = [] {
-		stringstream out_stream;
-		cin >> noskipws;
-		out_stream << cin.rdbuf();
-		return out_stream.str();
-	}();
-
-	if(out == ans) return {true, ""};
-
-	// Make lower case if needed.
-	if(not case_sensitive) {
-		for(auto& c : ans) c = tolower(c);
-		for(auto& c : out) c = tolower(c);
-		if(out == ans) return {true, "case"};
-	}
-
-	const auto& floatabs = float_absolute_tolerance;
-	const auto& floatrel = float_relative_tolerance;
-
-	if(space_change_sensitive and floatabs == 0 and floatrel == 0)
-		return {false, quick_diff(out, ans)};
-
-	// Split into tokens, depending on space_change_sensitive.
-	auto words = [](const string& st) {
-		stringstream s(st);
-		vector<string> words;
-		string w;
-		if(space_change_sensitive) {
-			s >> noskipws;
-			while(!s.eof()) {
-				if(s >> w) {
-					words.push_back(w);
-				} else {
-					s.clear();
-					assert(s.fail());
-					char c;
-					if(s >> c) {
-						assert(isspace(c));
-						words.emplace_back(1, c);
-					}
-				}
-			}
-		} else {
-			while(s >> w) words.push_back(w);
+// each token is either EOF, a single whitespace charachter or a string without any whitespace character
+struct token_view {
+	std::string_view token;
+
+	token_view() {}
+	token_view(std::string_view token) : token{token} {}
+
+	bool is_eof() const {
+		return token.empty();
+	}
+
+	std::optional<char> is_space() const {
+		if (token.size() != 1) return {};
+		if (not util::is_space(token[0])) return {};
+		return token[0];
+	}
+
+	std::optional<long double> is_float() const {
+		if (not util::is_float(token)) return {};
+		try {
+			//std::from_chars for floats is not widely supported ):
+			std::size_t pos = 0;
+			std::string tmp(token);
+			long double res = std::stold(tmp, &pos);
+			assert(pos == token.size());
+			assert(std::isfinite(res));
+			return res;
+		} catch(const std::out_of_range& /**/) {
+			// i dont know how to handle this... return +-inf?
```

Comment:
maybe just panic? Seems safer than possibly silently ignoring values.

On the other hand if teams submit rediculously long float values we probably shouldn't crash... so maybe just saying it's not a float and doing string comparison is fine. I think when we say 'output a float' we can expect something reasonable and reject unreasonable floats.

Reply:
panic would be ok for jury tokens (maybe?) but definitely not for team tokens... but yea I have no really god idea for this...

Reply:
we leave it as you did then